bindings/haskell/haskell-src/OpenDAL.hs

```diff
@@ -0,0 +1,87 @@
+-- Licensed to the Apache Software Foundation (ASF) under one
+-- or more contributor license agreements.  See the NOTICE file
+-- distributed with this work for additional information
+-- regarding copyright ownership.  The ASF licenses this file
+-- to you under the Apache License, Version 2.0 (the
+-- "License"); you may not use this file except in compliance
+-- with the License.  You may obtain a copy of the License at
+--
+--   http://www.apache.org/licenses/LICENSE-2.0
+--
+-- Unless required by applicable law or agreed to in writing,
+-- software distributed under the License is distributed on an
+-- "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+-- KIND, either express or implied.  See the License for the
+-- specific language governing permissions and limitations
+-- under the License.
+
+module OpenDAL (
+  Operator,
+  createOp,
+  readOp,
+  writeOp,
+) where
+
+import Data.ByteString (ByteString)
+import qualified Data.ByteString as BS
+import Data.HashMap.Strict (HashMap)
+import qualified Data.HashMap.Strict as HashMap
+import Foreign
+import Foreign.C.String
+import OpenDAL.FFI
```

Comment:
By providing a langauge binding, I expect users can work with it natively. For example, in python, users can just:

```py
import opendal

op = opendal.Operator("fs")
op.read("path/to/file")
```

Is it possible that we provide native hashell API instead? Take `amazonka-s3` as an example:

```hs
{-# LANGUAGE DataKinds #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE OverloadedLabels #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE TypeApplications #-}
{-# LANGUAGE ViewPatterns #-}

module S3 where

import Amazonka hiding (length)
import Amazonka.S3
import Data.ByteString (ByteString)
import Data.Conduit
import qualified Data.Conduit.Binary as CB
import qualified Data.Conduit.List as CL
import qualified Data.Foldable as Fold
import Data.Generics.Labels ()
import Data.Maybe (fromMaybe)
import Data.Text (Text)
import qualified Data.Text as Text
import qualified Data.Text.IO as Text
import Data.Time
import System.IO

getFile ::
  -- | Region to operate in.
  Region ->
  BucketName ->
  -- | The source object key.
  ObjectKey ->
  -- | The destination file to save as.
  FilePath ->
  IO ()
getFile reg b k f = do
  lgr <- newLogger Debug stdout
  env <- newEnv discover <&> set #logger lgr . set #region reg

  runResourceT $ do
    rs <- send env (newGetObject b k)
    view #body rs `sinkBody` CB.sinkFile f
    say $
      "Successfully Download: "
        <> b ^. _BucketName
        <> " - "
        <> k ^. _ObjectKey
        <> " to "
        <> Text.pack f
```

Reply:
We can encapsulate the `Operator` inside a `monad` to provide better error handling and avoid passing the Operator explicitly to functions.

The API will be like this.
```haskell
type OpenDAL a = ExceptT OpenDALError (ReaderT Operator IO) a

readOp :: String -> OpenDAL ByteString
writeOp :: String -> ByteString -> OpenDAL ()
```
And this is an example of using it.

```haskell
printFileContent :: FilePath -> OpenDAL ()
printFileContent path = do
  content <- readOp path
  liftIO $ putStrLn $ "File content:\n" ++ (BS.unpack content)
  
main :: IO ()
main = do
  let scheme = "fs"
      config = fromList [("root", "/")]
  result <- createOp scheme config
  case result of
    Right operator -> do
      let filePath = "/tmp/opendal/source/input.txt"
      actionResult <- runReaderT (runExceptT $ printFileContent filePath) operator
      case actionResult of
        Right () -> putStrLn "File content printed successfully."
        Left errMsg -> putStrLn $ "Error: " ++ show errMsg
    Left errMsg -> putStrLn $ "Failed to create operator: " ++ errMsg
```

If users want to ignore error handling, they can write like this.

```haskell
main :: IO ()
main = do
  let scheme = "fs"
      config = fromList [("root", "/")]
      filePath = "/tmp/opendal/source/input.txt"
  Right result <- createOp scheme config
  runReaderT (runExceptT $ printFileContent filePath) operator
```

Reply:
So we export the `FFI` here is for error handling?

Reply:
Haskell requires all compiled modules must be listed in .cabal file. We can move `OpenDAL.FFI` from `exposed-modules` to `other-modules` so that it will become invisible.

Reply:
> Haskell requires all compiled modules must be listed in .cabal file. We can move `OpenDAL.FFI` from `exposed-modules` to `other-modules` so that it will become invisible.

Nice!